### PR TITLE
refactor the phalcon/config.zep for issue #12613

### PR DESCRIPTION
**Pull request by qzfzz**
Hello!

* Type: code quality
* Link to issue:  https://github.com/phalcon/cphalcon/issues/12613
 #12613	

Small description of change:
The phalcon/config.zep has not fully implemented the ArrayAccess interface right now, so I refactor the phalcon/config.zep for fully implementing the interface!

Thanks

### CODE REVIEW

**Comment by Jurigag** [comment]
Update changelog, also we don't accept PRs to master branch.

File: phalcon/config.zep [hunk 1]
```diff
@@ -58,7 +59,7 @@ class Config implements \ArrayAccess, \Countable
 		var key, value;
 
 		for key, value in arrayConfig {
-			this->offsetSet(key, value);
```
Comment (cottton) [inline phalcon/config.zep]: would suggest to leave `this->offsetSet(key, value);`

File: phalcon/config.zep [hunk 2]
```diff
@@ -127,9 +131,9 @@ class Config implements \ArrayAccess, \Countable
 		let index = strval(index);
 
 		if typeof value === "array" {
-			let this->{index} = new self(value);
+			let this->arrayConfig[index] = value;
```
Comment (cottton) [inline phalcon/config.zep]: should be `let this->arrayConfig[index] = new self(value);`

File: phalcon/config.zep [hunk 3]
```diff
@@ -49,7 +49,8 @@ use Phalcon\Config\Exception;
  */
 class Config implements \ArrayAccess, \Countable
 {
-
+	protected arrayConfig = [];
```
Comment (cottton) [inline phalcon/config.zep]: suggest protected data = [];

File: phalcon/config.zep [hunk 4]
```diff
@@ -74,8 +75,7 @@ class Config implements \ArrayAccess, \Countable
 	public function offsetExists(var index) -> boolean
 	{
 		let index = strval(index);
-
-		return isset this->{index};
+		return this->arrayConfig[index];
```
Comment (cottton) [inline phalcon/config.zep]: return **isset** this->arrayConfig[index];

**Comment by qzfzz** [comment]
Hello! Thank you for your patience. This is the first time I contributing source code for phalcon, and I have correct the bugs for refactor the phalcon/config.zep.

**Comment by Jurigag** [comment]
Good but we don't accept PRs to master branch. Master is for latest stable version. Send your pr to 3.0.x and update changelog.

**Comment by qzfzz** [comment]
Thank you for your attention, I will create new pr to 3.0.x soon!


原始邮件
发件人:Wojciech Ślawskinotifications@github.com
收件人:phalcon/cphalconcphalcon@noreply.github.com
抄送:qzfzzqzfzz@163.com; Authorauthor@noreply.github.com
发送时间:2017年2月11日(周六) 22:07
主题:Re: [phalcon/cphalcon] refactor the phalcon/config.zep for issue#12613 (#12615)


Good but we don't accept PRs to master branch. Master is for latest stable version. Send your pr to 3.0.x and update changelog.

**Comment by cottton** [comment]
Would suggest another change btw.
Code: https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L188

     // updated: value instanceof Config &&
    if  value instanceof Config && method_exists(value, "toArray") {
        let arrayConfig[key] = value->toArray();
    } else {
        let arrayConfig[key] = value;
    }

**Reason**: should only convert to array if is instance of Config.
F.e. we got a class Car and Car has the method `::toArray`
then the current code would convert all the Car objects to array.
But user expected the **value** (object) car he previously set.

**Comment by Jurigag** [comment]
Then just instanceof check will be enough.

**Comment by cottton** [comment]
Indeed.

**Comment by cottton** [comment]
Another problem may exists at https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L93
If value of `this->{index}` is `null` then `defaultValue` will be returned.
User may expected `null` (offset actually exists).

Dunno how to check if property exists in zephir but:

    public function get(var index, var defaultValue = null) -> var
	{
        let index = strval(index);

        if property_exists(this, index) // property_exists ?
            return this->{index}
		}

		return defaultValue;
	}

---------------------

The same problem at `::offsetExists` https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L74

If offset `index` exists and is `null` it should return true.
Otherwise the method would have be named `offsetIsset`

**Comment by qzfzz** [comment]
Thank you so much for your advise.


Have a nice day!




原始邮件
发件人:cotttonnotifications@github.com
收件人:phalcon/cphalconcphalcon@noreply.github.com
抄送:qzfzzqzfzz@163.com; Authorauthor@noreply.github.com
发送时间:2017年2月13日(周一) 22:12
主题:Re: [phalcon/cphalcon] refactor the phalcon/config.zep for issue#12613 (#12615)


Another problem may exists at https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L93
 If value of this-{index} is null then defaultValue will be returned.
 User may expected null (offset actually exists).
Dunno how to check if property exists in zephir but:
public function get(var index, var defaultValue = null) - var { let index = strval(index); if property_exists(this, index) // property_exists ? return this-{index} } return defaultValue; }
The same problem at ::offsetExists https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L74
If offset index exists and is null it should return true.
 Otherwise the method would have be named offsetIsset

**Comment by qzfzz** [comment]
Thank you so much, and I have fixed the bugs!


原始邮件
发件人:cotttonnotifications@github.com
收件人:phalcon/cphalconcphalcon@noreply.github.com
抄送:qzfzzqzfzz@163.com; Authorauthor@noreply.github.com
发送时间:2017年2月13日(周一) 22:12
主题:Re: [phalcon/cphalcon] refactor the phalcon/config.zep for issue#12613 (#12615)


Another problem may exists at https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L93
 If value of this-{index} is null then defaultValue will be returned.
 User may expected null (offset actually exists).
Dunno how to check if property exists in zephir but:
public function get(var index, var defaultValue = null) - var { let index = strval(index); if property_exists(this, index) // property_exists ? return this-{index} } return defaultValue; }
The same problem at ::offsetExists https://github.com/phalcon/cphalcon/blob/3.0.x/phalcon/config.zep#L74
If offset index exists and is null it should return true.
 Otherwise the method would have be named offsetIsset

**Comment by cottton** [comment]
Question: 
`\Phalcon\Builder\Options` extends `Phalcon\Config`.
See https://github.com/phalcon/phalcon-devtools/blob/3.0.x/scripts/Phalcon/Builder/Options.php

If in class `Options` (or any other extending class) a value is set like `$this->{$key}` then the setter [(offsetSet)](https://github.com/qzfzz/cphalcon/blob/7d4c5cf0cee8eab7c0877b044e114712e9e68474/phalcon/config.zep#L130) gets called, right?

**Comment by Jurigag** [comment]
No. offsetSet is method from ArrayAccess interface. It's making possible to do something like $this[$key] = 'xyz' and it will set property called $key with value 'xyz'.

http://php.net/manual/en/class.arrayaccess.php

**Comment by cottton** [comment]
Ok. In the current fix we use `protected $data` that hold the data.
In PHP if we extend `Phalcon\Config` and use `$his->{$key} = $value;` - do we have an conflict?

**Comment by Jurigag** [comment]
If someone will use `$this['data'] = $value` then yes.

**Comment by Jurigag** [comment]
Well tbh i don't see offsetSet changed method here. Why is that?

**Comment by Jurigag** [comment]
Oh i guess code is changed so there will be no conflict.

**Comment by Jurigag** [comment]
Can you close this since it's to master repo?

**Pull request closed without merging**